plugins/Rockets/SnapshotTask.cpp

```diff
@@ -0,0 +1,188 @@
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <brayns/common/tasks/Task.h>
+
+#include "ImageGenerator.h"
+#include <brayns/common/utils/utils.h>
+#include <brayns/engine/Camera.h>
+#include <brayns/engine/Engine.h>
+#include <brayns/engine/FrameBuffer.h>
+#include <brayns/engine/Renderer.h>
+#include <brayns/engine/Scene.h>
+
+#include <brayns/parameters/ParametersManager.h>
+
+#include "SnapshotTask.h"
+
+namespace brayns
+{
+SnapshotFunctor::SnapshotFunctor(Engine &engine, SnapshotParams &&params,
+                                 ImageGenerator &imageGenerator)
+    : _params(std::move(params))
+    , _imageGenerator(imageGenerator)
+    , _engine(engine)
+{
+    if (_params.animParams == nullptr)
+    {
+        _params.animParams = std::make_unique<AnimationParameters>(
+            engine.getParametersManager().getAnimationParameters());
+    }
+
+    if (_params.geometryParams == nullptr)
+    {
+        _params.geometryParams = std::make_unique<GeometryParameters>(
+            engine.getParametersManager().getGeometryParameters());
+    }
+
+    if (_params.renderingParams == nullptr)
+    {
+        _params.renderingParams = std::make_unique<RenderingParameters>(
+            engine.getParametersManager().getRenderingParameters());
+    }
+
+    if (_params.volumeParams == nullptr)
+    {
+        _params.volumeParams = std::make_unique<VolumeParameters>(
+            engine.getParametersManager().getVolumeParameters());
+    }
+
+    if (_engine.supportsConcurrentSnapshots())
+    {
+        _camera = engine.createCamera();
+        _scene =
+            engine.createScene(*_params.animParams, *_params.geometryParams,
+                               *_params.volumeParams);
+
+        _renderer = engine.createRenderer(*_params.animParams,
+                                          *_params.renderingParams);
+
+        const auto &renderer = engine.getRenderer();
+        _renderer->setCurrentType(renderer.getCurrentType());
+        _renderer->clonePropertiesFrom(renderer);
+        if (_params.camera)
+        {
+            *_camera = *_params.camera;
+            _camera->setCurrentType(engine.getCamera().getCurrentType());
+            _camera->clonePropertiesFrom(engine.getCamera());
+        }
+        else
+            *_camera = engine.getCamera();
+
+        _scene->copyFrom(engine.getScene());
+    }
+    else
+    {
+        _camera = engine.getCameraPtr();
+        _renderer = engine.getRendererPtr();
+    }
+}
+
+ImageGenerator::ImageBase64 SnapshotFunctor::_renderSnapshot(
+    const size_t numAccumulations)
+{
+    const auto nameShort =
+        _params.name.empty() ? "" : " " + shortenString(_params.name);
```

Comment:
Type deduction rules must be really complicated if this can successfully infer that nameShort is std::string...